W-20064735 grpc 1.0 (#3043)
mulesoft/docs-connectors: grpc/1.0/modules/ROOT/pages/grpc-connector-reference.adoc

= gRPC Connector 1.0 Reference

Anypoint Connector for gRPC (grPC Connector) enables you to connect to the gRPC API.

== Configurations
---
[[Client]]
=== Client

GrpcClientConfiguration class.

==== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required

> |Name | String | The name for this configuration. Connectors reference the configuration with this name. | | x

Reviewer: ```suggestion
|Name | String | The name for this configuration. Connectors reference the configuration with this name. The identifier of this element used to reference it in other components. | | x
```

| Connection a| * <<Client_NoAuth, No Auth>> 
* <<Client_OauthClientCredentials, Oauth Client Credentials>> 
 | The connection types to provide to this configuration. | | x

> | Name a| String | The identifier of this element used to reference it in other components. |  | x

Reviewer: ```suggestion
```

Author: This is in our other connector docs too, it's intentional and from the generated reference/UI

| Client Secret a| String | The OAuth client secret as registered with the service provider. |  | x
| Token Url a| String | The service provider's token endpoint URL. |  `\https://example.com/oauth2/token` | 
| Scopes a| String | The OAuth scopes to be requested during the dance. If not provided, it defaults to those in the annotation. |  | 
| Auth Server Tls Context a| <<Tls>> | References a TLS config to use to do HTTP request during the OAuth dance. |  | 
| Auth Server Proxy Config a| Any | References the proxy configuration which through the HTTP request during the OAuth dance. |  | 
| Object Store a| String | A reference to the object store that should be used to store each resource owner id's data. If not specified, runtime will automatically provision the default one. |  | 
|===

== Supported Operations
* <<BidirectionalStreaming>> 
* <<ClientStreaming>> 
* <<ServerStreamingCallWithDelegation>> 
* <<ServerStreamingCallWithResponse>> 
* <<UnaryCall>> 
* <<Unauthorize>> 

== Associated Sources
* <<BidiStreamingListener>> 
* <<ServerStreamingDelegatedListener>> 
* <<ServerStreamingSubscriberSource>> 

---
[[Config]]
=== Config

Default configuration.

==== Parameters

> [%header%autowidth.spread]
> |===
> | Name | Type | Description | Default Value | Required
> |Name | String | The name for this configuration. Connectors reference the configuration with this name. | | x
> | Name a| String | The identifier of this element used to reference it in other components. |  | x

Reviewer: ```suggestion
|Name | String | The name for this configuration. Connectors reference the configuration with this name. The identifier of this element used to reference it in other components. |  | x
```

|===

== Operations

[[BidirectionalStreaming]]
== Bidirectional Streaming Call
`<grpc:bidirectional-streaming>`

> Connect to a Bidirectional Streaming RPC. This operation initiates a request, and its responses are collected by the corresponding Bidirectional Streaming Listener. The Bidirectional Streaming Call and Bidirectional Streaming Listener are mapped to each other via the Routing Id. This operation can be used to initiate client streaming, send requests, and also close the client streaming. Each of these client steps can be mapped using a common request ID.

Reviewer: ```suggestion
Connect to a Bidirectional Streaming RPC. This operation initiates a request, and its responses are collected by the corresponding Bidirectional Streaming Listener. The Bidirectional Streaming Call and Bidirectional Streaming Listener are mapped to each other via the Routing Id. This operation can be used to initiate the call, opening the client streaming, sending requests, and also closing the client streaming. Each of these client steps can be mapped using a common request ID.
```

Author: Made this change in separate commit
https://github.com/mulesoft/docs-connectors/pull/3043/commits/d6aec1611c96b1c2cde6943052f78a2e3d7f08a4

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Method a| String | The gRPC method name to invoke. |  | x
| Routing Id a| String | Identifier to map this call with its corresponding listener. |  | 
| Request Id a| String | Identifier to map related client streaming steps. |  | 
| Content a| Binary | The request payload to send to the gRPC server. |  | 
| End Streaming a| Boolean | Indicates whether to close the client streaming. |  false | 
| Request Headers a| Object | Custom headers to include in the gRPC request. |  | 
| Config Ref a| ConfigurationProvider | The name of the configuration to use to execute this component. |  | x

> | Target Variable a| String | Name of the variable that stores operation's output. |  | 
> | Target Value a| String | An expression to evaluate against the operation's output and store the expression outcome in the target variable. |  #[payload] | 

Reviewer: These are for any operation. I would not document this in this page. It is documented elsewhere -> https://docs.mulesoft.com/mule-runtime/latest/target-variables 
```suggestion
```

Author: This is in our other connector docs too, it's intentional and from the generated reference/UI

> | Error Mappings a| Array of <<ErrorMapping>> | Set of error mappings. |  | 
> | Reconnection Strategy a| * <<Reconnect>>
> * <<ReconnectForever>> | A retry strategy in case of connectivity errors. |  | 

Reviewer: Also not specific to the connector but standard functioning of mule operations.
```suggestion
```

* GRPC:RETRY_EXHAUSTED 

[[ClientStreaming]]
== Client Streaming Call
`<grpc:client-streaming>`

Connect to a Client Streaming RPC. This single operation can be used to initiate streaming, send requests, and close the client streaming request. Each of these steps can be mapped using a common request ID.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Method a| String | The gRPC method name to invoke. |  | x
| Request Id a| String | Identifier to map related client streaming steps. |  | 
| Content a| Binary | The request payload to send to the gRPC server. |  | 
| End Streaming a| Boolean | Indicates whether to close the client streaming. |  false | 
| Request Headers a| Object | Custom headers to include in the gRPC request. |  | 
| Config Ref a| ConfigurationProvider | The name of the configuration to use to execute this component. |  | x
| Streaming Strategy a| * <<RepeatableInMemoryStream>>
* <<RepeatableFileStoreStream>>
* non-repeatable-stream | Configures how Mule processes streams. Repeatable streams are the default behavior. |  | 

> | Target Variable a| String | Name of the variable that stores operation's output. |  | 
> | Target Value a| String | An expression to evaluate against the operation's output and store the expression outcome in the target variable. |  #[payload] | 
> | Error Mappings a| Array of <<ErrorMapping>> | Set of error mappings. |  | 
> | Reconnection Strategy a| * <<Reconnect>>
> * <<ReconnectForever>> | A retry strategy in case of connectivity errors. |  | 

Reviewer: ```suggestion
```

[%autowidth.spread]
|===
|Type |Binary
| Attributes Type a| <<ClientStreamingAttributes>>
|===

=== For Configurations

* <<Client>> 

=== Throws

* GRPC:CONNECTIVITY 
* GRPC:RETRY_EXHAUSTED 

[[ServerStreamingCallWithDelegation]]
== Server Streaming Delegator
`<grpc:server-streaming-call-with-delegation>`

Connect to a Server Streaming RPC. This operation initiates a request, and its responses are collected by the corresponding Server Streaming Delegated Listener. The Server Streaming Delegator and Server Streaming Delegated Listener are mapped to each other via the routing ID.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Method a| String | The gRPC method name to invoke. |  | x
| Routing Id a| String | Identifier to map this call with its corresponding listener. |  | x
| Content a| Binary | The request payload to send to the gRPC server. |  #[payload] | 
| Request Headers a| Object | Custom headers to include in the gRPC request. |  | 
| Config Ref a| ConfigurationProvider | The name of the configuration to use to execute this component. |  | x

> | Target Variable a| String | Name of the variable that stores operation's output. |  | 
> | Target Value a| String | An expression to evaluate against the operation's output and store the expression outcome in the target variable. |  #[payload] | 
> | Error Mappings a| Array of <<ErrorMapping>> | Set of error mappings. |  | 
> | Reconnection Strategy a| * <<Reconnect>>
> * <<ReconnectForever>> | A retry strategy in case of connectivity errors. |  | 

Reviewer: ```suggestion
```

|===

=== Output

[%autowidth.spread]
|===
|Type |String
|===

=== For Configurations

* <<Client>> 

=== Throws

* GRPC:CONNECTIVITY 
* GRPC:RETRY_EXHAUSTED 

[[ServerStreamingCallWithResponse]]
== Server Streaming Iterator
`<grpc:server-streaming-call-with-response>`

Connect to a Server Streaming RPC. This operation provides a StreamedResponseIterable as output, which can be used to iterate over server responses.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Method a| String | The gRPC method name to invoke. |  | x
| Content a| Binary | The request payload to send to the gRPC server. |  #[payload] | 
| Request Headers a| Object | Custom headers to include in the gRPC request. |  | 
| Output Mime Type a| String | The mime type of the payload that this operation outputs. |  | 
| Output Encoding a| String | The encoding of the payload that this operation outputs. |  | 
| Config Ref a| ConfigurationProvider | The name of the configuration to use to execute this component. |  | x

> | Target Variable a| String | Name of the variable that stores operation's output. |  | 
> | Target Value a| String | An expression to evaluate against the operation's output and store the expression outcome in the target variable. |  #[payload] | 
> | Error Mappings a| Array of <<ErrorMapping>> | Set of error mappings. |  | 
> | Reconnection Strategy a| * <<Reconnect>>
> * <<ReconnectForever>> | A retry strategy in case of connectivity errors. |  | 

Reviewer: ```suggestion
```

= gRPC Connector 1.0 Reference

> Anypoint Connector for gRPC (grPC Connector) enables you to connect to the gRPC API.

Reviewer 2: ```suggestion
Anypoint Connector for gRPC (gRPC Connector) enables you to connect to the gRPC API.
```
